Added support for saving and loading non ASCII chars in corpus and vocab

IssacXid opened the pull request:
Made the changes as per issue #79

xhluca commented on bm25s/__init__.py:
```diff
@@ -876,8 +876,8 @@ def save(
-        with open(vocab_path, "w") as f:
-            f.write(json_functions.dumps(self.vocab_dict))
+        with open(vocab_path, "wt", encoding='utf-8') as f:  
+            json.dump(self.vocab_dict, f, ensure_ascii=False)
```
Could you make the changes directly to json_functions, such that it accepts a kwargs? This way we can keep using orjson when possible.

I'm also curious if orjson supports ensure_ascii... Do you have any experience?

IssacXid replied:
@xhluca In orjson, the default behavior already aligns with ensure_ascii=False in Python's built-in json.dump and json.dumps. I've modified the json_functions file to take kwargs also. I've tested it with non ASCII characters for which it was throwing errors earlier. Please review.

xhluca reviewed (changes requested):
See comment, also could you add the new test case from the issue?

IssacXid commented:
Regarding the test files, isn't adding some non ASCII texts in the corpus sufficient?

xhluca commented:
Thanks for the 2nd commit, unfortunately seems like there's some issue causing the tests to fail: https://github.com/xhluca/bm25s/actions/runs/11994683583/job/33448614792#step:6:1

Perhaps running the tests locally could help diagnose the issue?

IssacXid commented:
Sorry, I didn't quite know how to test all the test files at once, so only checked  the core/test_save_load.py. Now, I've run both core and comparison tests using `python -m unittest tests/(core or comparison)/test_*.py` cmd, it ran without error. Please review. Thanks

xhluca commented:
Can you elaborate on the changes to the test file: tests/core/test_tokenizer_misc.py ?

IssacXid commented:
> Can you elaborate on the changes to the test file: tests/core/test_tokenizer_misc.py ?

In my local, I saw that test_tokenizer_misc.py was failing as the ids were not matching the results. In my case, it was coming up to be [2,0,3]. Then I checked with the corpus and query:
```
corpus = [
            "a cat is a feline and likes to purr",
            "a dog is the human's best friend and loves to play",
            "a bird is a beautiful animal that can fly",
            "a fish is a creature that lives in water and swims",
        ]
query = "What is a fly?"
```
Thought that the id 2 `"a bird is a beautiful animal that can fly"` should be at least ranked first, which was also the result coming in my case. So, I edited the assertion.

IssacXid commented:
Yes, github is allowing me to see the test results. I've changed back the assertion in test_tokenizer_misc.py. I also added the test as #79, which is passing in my local. In case, this doesn't pass in the github actions, let's please connect over a webcall.

xhluca commented:
Looks good!

IssacXid commented:
Hi @xhluca, updated this [hotfix/save-load-extended-character-corpora ](https://github.com/IssacXid/bm25s/tree/hotfix/save-load-extended-character-corpora). Missed to add ensure-ascii=False while saving corpus. It is causing `invalid escape sequence` failure while retrieving along with corpus. Should I create another PR or can you re-open this one?

xhluca commented:
Please create a new one
